### DISCUSSION TITLE
LDAP auth wrong credentials management with hook

### DISCUSSION THREAD

**Discussion by mhdaouas** [directus/directus#9055] [Archive]
In case of LDAP auth, which hook should I use to check user existence and his credentials before allowing him to access the app? What is the proper way to prevent the user from authenticating in case his credentials are wrong? This is not documented.

I tried with the auth.login.before hook to prevent user access with "throw new InvalidPayloadException" but this method crashes the app (called in case of wrong credentials).

I'm using ldapjs lib FYI.

**Comment by aidenfoxx** [comment]
If you can wait a week, official LDAP support should be coming. 👍

**Reply by mhdaouas** [reply]
I already implemented it in the auth.login.before hook but I have no idea if it is the right hook to start with and also no idea how to proceed in case of wrong credentials without crashing the app. In case of valid credentials, I get access to the app as expected without any problem.

**Comment by joselcvarela** [comment]
Hello @mhdaouas 
As you may know, there was a PR which handles LDAP auth https://github.com/directus/directus/pull/6821
Although, that was closed.
I think the main reason was to support modular authentication https://github.com/directus/directus/pull/6942 in order to create `drivers` for each use case.
For example, for OpenID and OAuth https://github.com/directus/directus/pull/8660

That being said, if you want to implement a custom way in the meantime, you can use custom hook `auth.login.before` and `InvalidCredentialsException` exception, instead of `InvalidPayloadException`.

By the way, can you explain this part:
> but this method crashes the app (called in case of wrong credentials).
Maybe there is some bug. I think the API should do not crash in any case.

**Reply by aidenfoxx** [reply]
@joselcvarela The LDAP code lives here https://github.com/aidenfoxx/directus/pull/2

It depended on the new OAuth ticket being merged since there was some changes made the the core authentication.

There are some merge conflicts with the LDAP branch, but once I clear them I'll be adding the PR to directus (probably tonight).

**Reply by joselcvarela** [reply]
Thank you @aidenfoxx 
I went to your fork but couldn't I don't know why I didn't find it 🤨

**Reply by aidenfoxx** [reply]
I got you fam https://github.com/directus/directus/pull/9074

**Reply by mhdaouas** [reply]
> By the way, can you explain this part:
> 
> > but this method crashes the app (called in case of wrong credentials).
> > Maybe there is some bug. I think the API should do not crash in any case.

From what I understood, in case of successful login, I just have to return the payload but for unsuccessful login I didn't know what to return so I tried throwing an InvalidPayloadException but this actually crashes Directus. I haven't tried yet with InvalidCredentialsException. I will let you know what happens.

**Reply by mhdaouas** [reply]
@aidenfoxx in my case, upon the first login from a specific user, I need to check if he is created in the directus db, if not, he has to be created in the db. When he logs the next time, only his info is updated (retrieved from LDAP server). Does your implementation covers this scenario or should I implement it manually?

**Reply by aidenfoxx** [reply]
@mhdaouas This exists is in my implementation. On first authentication, the user is created. On subsequent logins only their role is updated.

**Reply by mhdaouas** [reply]
Perfect :D
